extension/content/components/recipes/form/arguments/MessagingExperiment.js

```diff
@@ -0,0 +1,221 @@
+import JsonEditor from "devtools/components/common/JsonEditor";
+
+export default function MessagingExperiment() {
+  return (
+    <>
+      <InputField label="Experiment Slug" name="slug" />
+
+      <ToggleField label="Prevent New Enrollment" name="isEnrollmentPaused">
+        Prevents new users from joining the experiment cohort.
+        <br />
+        Existing users will remain in the experiment.
+      </ToggleField>
+
+      <FormGroup>
+        <Row>
+          <Col xs={24}>
+            <Branches />
+          </Col>
+        </Row>
+      </FormGroup>
+    </>
+  );
+}
+
+function Branches() {
+  const [branchesUpdated, setBranchesUpdated] = React.useState(Date.now());
+  const data = useRecipeDetailsData();
+  const dispatch = useRecipeDetailsDispatch();
+
+  const handleClickAddBranch = () => {
+    const newBranch = { ratio: 1, slug: "", value: {}, groups: [] };
+
+    dispatch({
+      type: ACTION_UPDATE_DATA,
+      data: {
+        ...data,
+        arguments: {
+          ...data.arguments,
+          branches: [...data.arguments.branches, newBranch],
+        },
+      },
+    });
+  };
+
+  let branchesList = (
+    <>
+      <HelpBlock>There are no branches.</HelpBlock>
+      <Divider />
+    </>
+  );
+  if (data.arguments.branches && data.arguments.branches.length) {
+    branchesList = data.arguments.branches.map((branch, index) => (
+      <Branch
+        key={index}
+        branchesUpdated={branchesUpdated}
+        index={index}
+        setBranchesUpdated={setBranchesUpdated}
+      />
+    ));
+  }
+
+  return (
+    <FormGroup>
+      <ControlLabel>Branches</ControlLabel>
+      <Panel bordered>
+        {branchesList}
+        <IconButton
+          icon={<Icon icon="plus-circle" />}
+          onClick={handleClickAddBranch}
+        >
+          Add Branch
+        </IconButton>
+      </Panel>
+    </FormGroup>
+  );
+}
+
+function Branch({ index, branchesUpdated, setBranchesUpdated }) {
+  const data = useRecipeDetailsData();
+  const dispatch = useRecipeDetailsDispatch();
+  const branch = data.arguments.branches[index];
+  const { branches } = data.arguments;
+
+  const handleClickDelete = () => {
+    setBranchesUpdated(Date.now());
+    dispatch({
+      type: ACTION_UPDATE_DATA,
+      data: {
+        ...data,
+        arguments: {
+          ...data.arguments,
+          branches: branches.filter((b, i) => i !== index),
+        },
+      },
+    });
+  };
+
+  const handleChange = (name, cb = (v) => v) => {
+    return (value) => {
+      dispatch({
+        type: ACTION_UPDATE_DATA,
+        data: {
+          ...data,
+          arguments: {
+            ...data.arguments,
+            branches: branches.map((b, i) => {
+              if (i === index) {
+                return { ...b, [name]: cb(value) };
+              }
+
+              return b;
+            }),
+          },
+        },
+      });
+    };
+  };
+
+  const parseNumericInput = (value) => {
+    if (!value) {
+      return "";
```

Comment:
So this returns a number sometimes and a string other times? Should it be `null` instead of `""`?

Reply:
null causes issues where it switches between controlled and uncontrolled for some reason. i believe that's related to how rsuite passes value through

Reply:
Ah, I can see how that happens. Thanks for the explanation!